tests/Unit/TolerantVariableReferenceFinderTest.php

```diff
@@ -0,0 +1,201 @@
+            $uri,
+        ];
+
+        yield 'scope: inside class method: select argument' => [
+            '<?php class C1 { function M1(<sr>$va<>r1) { <sr>$var1 = 5; $var2 = <sr>$var1 + 10; } }',
+            $uri,
+        ];
+
+        yield 'scope: inside class method: inside anonumous function + use, click inside' => [
+            '<?php class C1 { function M1() { <sr>$var1 = 10; $f = function() use (<sr>$var1) { <sr>$v<>ar1 = 5; $var2 = <sr>$var1 + 10; } } }',
+            $uri,
+        ];
+
+        yield 'scope: inside class method: inside anonumous function + use, click outside' => [
+            '<?php class C1 { function M1() { <sr>$v<>ar1 = 10; $f = function() use (<sr>$var1) { <sr>$var1 = 5; $var2 = <sr>$var1 + 10; } } }',
+            $uri,
+        ];
+
+        yield 'scope: inside class method: inside anonumous function + use, click in use' => [
+            '<?php class C1 { function M1() { <sr>$var1 = 10; $f = function() use (<sr>$v<>ar1) { <sr>$var1 = 5; $var2 = <sr>$var1 + 10; } } }',
+            $uri,
+        ];
+
+        yield 'scope: inside class method: inside anonumous function (no use), click inside' => [
+            '<?php class C1 { function M1() { $var1 = 10; $f = function(<sr>$var1) { <sr>$v<>ar1 = 5; $var2 = <sr>$var1 + 10; } } }',
+            $uri,
+        ];
+
+        yield 'scope: inside class method: inside anonumous function (no use), click outside' => [
+            '<?php class C1 { function M1() { <sr>$v<>ar1 = 10; $f = function($var1) { $var1 = 5; $var2 = $var1 + 10; } } }',
+            $uri,
+        ];
+
+        yield 'scope: inside class method: inside anonumous class method' => [
+            '<?php '.
+                'class C1 { function M1() { '.
+                '$var = 1;'.
+                '$c = new class { function IM() { <sr>$v<>ar = 1; } } '.
+                ' } }',
+            $uri,
+        ];
+    }
+
+    private static function offsetsFromSource(string $source, ?string $uri): array
+    {
+        $textDocumentUri = $uri !== null ? TextDocumentUri::fromString($uri) : null;
+        $results = preg_split("/(<>|<sr>|<mr>)/u", $source, null, PREG_SPLIT_DELIM_CAPTURE | PREG_SPLIT_NO_EMPTY);
+        
+        $referenceLocations = [];
+        $selectionOffset = null;
+
+        if (is_array($results)) {
+            $newSource = "";
+            $offset = 0;
+            foreach ($results as $result) {
+                if ($result == "<>") {
+                    $selectionOffset = $offset;
+                } elseif ($result == "<sr>") {
+                    $referenceLocations[] = PotentialLocation::surely(
+                        new Location($textDocumentUri, ByteOffset::fromInt($offset))
+                    );
+                } elseif ($result == "<mr>") {
+                    $referenceLocations[] = PotentialLocation::maybe(
+                        new Location($textDocumentUri, ByteOffset::fromInt($offset))
+                    );
+                } else {
```

Comment:
I would not use `else` (ever almost), instead you can use `continue`:

```
if ($result === '<>) {
   // do something
   continue; // we're done
}

if ($result == '...') {
  // ...
}

// ...

throw new Exception('...');
```

you can also reducing the nesting here by removing the outer else:

```
if (!is_array($_)) {
    return [ ... ];
}

// rest of the method is not unnecessarily indented
```

Reply:
I choose them by gut feeling most of the time - whatever seems more readable in the current code, but I am happy to do whatever you like! Goona fix it.

Reply:
How do you like that version:
```php
            foreach ($results as $result) {
                if ($result == "<>") {
                    $selectionOffset = $offset;
                    continue;
                }
                
                if ($result == "<sr>") {
                    $referenceLocations[] = PotentialLocation::surely(
                        new Location($textDocumentUri, ByteOffset::fromInt($offset))
                    );
                    continue;
                } 
                if ($result == "<mr>") {
                    $referenceLocations[] = PotentialLocation::maybe(
                        new Location($textDocumentUri, ByteOffset::fromInt($offset))
                    );
                    continue;
                }
                
                $newSource .= $result;
                $offset += mb_strlen($result);
            }
```

Reply:
yep perfect :+1: